### PR TITLE
Deprivatising Data.Nat

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
A fresh version of pull request https://github.com/agda/agda-stdlib/pull/103

Moved `decTotalOrder` in `Data.Nat` to `≤-decTotalOrder` in `Data.Nat.Properties` and removed the proofs from private scope. As a consequence also had to move the module `≤-Reasoning` to `Data.Nat.Properties` as well. 

I feel that the module `≤-Reasoning` should really be moved to its own file, but that can always be done in another pull request after this is accepted.

Obviously `Data.Nat` is now virtually empty so, if there's an appetite for it, the contents of `Data.Nat.Base` can be moved into `Data.Nat` later on.

### CODE REVIEW

**Comment by andreasabel** [comment]
Er, I am confused.  Does `Data.Nat` export less now? If yes, why is that good?

> Obviously Data.Nat is now virtually empty so, if there's an appetite for it, the contents of Data.Nat.Base can be moved into Data.Nat later on.

The idea of `Data.Nat.Base` is to have a lean module to have the most essential definitions for natural numbers.  `Data.Nat` will then give you something more comprehensive, but be slow, as it depends on a lot of other modules.

**Comment by MatthewDaggitt** [comment]
Sorry I meant **selectively** move the contents of 'Data.Nat.Base' back into 'Data.Nat'. For example the things that don't seem essential to me are:

-  Anything to do with `_≤′_` and `_≤′′_`
- The `GeneralisedArithmetic` module
- The `Ordering` definition.

These are just suggestions for the future of course! The important thing is that we've made the ordering properties accessible and moved them out of `Data.Nat` as they didn't really belong there.  Is the pull request okay as is?

**Comment by andreasabel** [comment]
Just to confirm:  Since `Data.Nat` now exports less, user code may break because of things no longer being in scope that used to be in scope!?

**Comment by MatthewDaggitt** [comment]
Yes that's true. `Data.Nat` no longer exports `decTotalOrder` or `≤-reasoning`.

I think that `decTotalOrder`, if following the convention used in other parts of the library, is definitely in the wrong place in `Data.Nat` as the fact that `_≤_` is a decidable total order is a property, not an operation or relation on natural numbers, and therefore should be in `Data.Nat.Properties`. 

Furthermore `decTotalOrder` *must* be moved in order to carry out your suggestion in https://github.com/agda/agda-stdlib/pull/103 that the proofs `≤-refl` `≤-trans` etc be put in `Data.Nat.Properties` as it's not possible (nor desirable) to have the `Data.Nat` depend on `Data.Nat.Properties`.

TLDR: it does break backwards compatibility but `decTotalOrder`, as it stands, is in the wrong place and moving it is an unavoidable consequence of making the basic properties more easily available in the right place.

**Comment by andreasabel** [comment]
Ok.

**Comment by andreasabel** [comment]
Mmh, I merged, but already the Agda test-suite breaks...
I'd be great if you could run the `make test` of Agda with your library branch before a PR.

**Comment by MatthewDaggitt** [comment]
I'm so sorry... I don't know how I missed that... When I was making all the changes outside of `Data.Nat` I found all the other files that needing changing by running `make test` repeatedly until it didn't throw any more errors. After running `make test` the final time I must have missed `DivModUtils` out my `git add`. I'm sorry that's very sloppy of me. I'm away from my machine, otherwise I'd fix it myself. My apologies, once again.

**Comment by andreasabel** [comment]
No, problem, it was not hard to fix.  Just wanted to tell you to alert you to the general procedure, but you are already following it, so no problem.

**Comment by asr** [comment]
> I'd be great if you could run the `make test` of Agda with your library branch before a PR.

For what? If the PR's author find a problem running `make test` on Agda, he cannot fix the problem using a PR for the standard library. 

> After running `make test` the final time I must have missed `DivModUtils` out my `git add`.

I don't think so. I guess you didn't run `make test` on the Agda repository but in the library repository which is the correct procedure.

**Comment by andreasabel** [comment]
> For what? If the PR's author find a problem running make test on Agda, he cannot fix the problem using a PR for the standard library.

True, but he can alert about the expected breakage, and also get a bit more feeling about the backwards-compatibility of the change.

**Comment by MatthewDaggitt** [comment]
Oh, now I understand. It failed tests in the Agda repository not the standard library repository. It didn't even occur to me that the Agda repository might have dependencies on the standard library... 

I've downloaded the Agda repository but I can't seem to get the tests to locate my altered version of the standard library. I'm following the instructions in HACKING.md in the Agda repository and updated the file `/.gitmodules` to contain a URL to my own fork. I then run `make std-lib` but the tests still seem to be running the old version of the std-lib. Am I missing something? Do I have to feed it a specific commit in `/std-lib`?

In the std-lib `HACKING.md` file it might be worth adding  that any changes that break backwards compatibility need to be tested against the main Agda repository as well, and put a few pointers about how to do it. Might help well-meaning but essentially clueless people like me from disrupting everything again!

**Comment by asr** [comment]
> Do I have to feed it a specific commit in `/std-lib`?

It should work.

**Comment by MatthewDaggitt** [comment]
Hmm so my `.gitmodule` file is as follows:
```
[submodule "std-lib"]
	path = std-lib
	url = https://github.com/MatthewDaggitt/agda-stdlib
	branch = deprivatising-DataNatProperties
```
but when I run `make std-lib` it's pulling the `master` branch of my repository `https://github.com/MatthewDaggitt/agda-stdlib` rather than the `deprivatising-DataNatProperties` branch.

**Comment by MatthewDaggitt** [comment]
Okay copied across the branch manually and now I get the following error when running `make test` 
```
======================================================================
===================== Suite of successfull tests =====================
======================================================================
make[1]: Entering directory `XXX/agda-update/test/Common'
Bool.agda
make[1]: XXX/agda-update/dist-2.6.0/build/agda/agda: Command not found
make[1]: *** [Bool.test] Error 127
make[1]: Leaving directory `XXX/agda-update/test/Common'
make: *** [succeed] Error 2
```
where `XXX/agda-update` is the location of my copy of the main Agda repository.

**Comment by asr** [comment]
> `make[1]: XXX/agda-update/dist-2.6.0/build/agda/agda: Command not found`

`make test` uses the Agda binary generated by Cabal on the above directory. You need to run 

```
make install-bin
```

before running the test suite.
